### PR TITLE
Add type hints

### PR DESCRIPTION
**Pull request by jason20c**
For #66

### CODE REVIEW

File: src/confu/cli.py [hunk 1]
```diff
@@ -26,7 +34,7 @@ def option_name(path, delimiter="--"):
     return "--{}".format(delimiter.join(path).replace("_", "-"))
 
 
-def destination_name(path, delimiter="__"):
+def destination_name(path: list, delimiter: str = "__") -> str:
```
Comment (vegu) [inline src/confu/cli.py:37]: i think that should be `path: typing.List[str]` @jason20c

File: src/confu/cli.py [hunk 2]
```diff
@@ -42,7 +50,7 @@ def destination_name(path, delimiter="__"):
     return f"{delimiter.join(path)}"
 
 
-def default(value, path, defaults):
+def default(value: Any, path: list, defaults: dict | None) -> Any:
```
Comment (vegu) [inline src/confu/cli.py:53]: i think that should be path: typing.List[str] @jason20c

File: src/confu/schema/core.py [hunk 3]
```diff
@@ -3,12 +3,18 @@
...
 """
+from __future__ import annotations
 
 import collections.abc
 import configparser
 import copy
 import os
 from inspect import isclass
+from typing import Any, Callable
+from typing import Dict as typing_Dict
```
Comment (vegu) [inline src/confu/schema/core.py:14]: i think in these cases where typing clashes with confu schema types (confu.Dict vs typing.Dict) we should probably handle it via
```
import typing
```

and then use `typing.Dict` for the type hints

@jason20c

File: src/confu/schema/inet.py [hunk 4]
```diff
@@ -22,7 +26,7 @@ class Email(Str):
     Describes an email address
     """
 
-    def validate(self, value, path, **kwargs):
+    def validate(self, value: str | None, path: list, **kwargs: Any) -> str | None:
```
Comment (vegu) [inline src/confu/schema/inet.py:29]: i think that should be path: typing.List[str] @jason20c

File: src/confu/schema/inet.py [hunk 5]
```diff
@@ -44,11 +48,11 @@ class Url(Str):
     Describes a URL
     """
 
-    def __init__(self, name="", **kwargs):
+    def __init__(self, name: str = "", **kwargs: Any) -> None:
         super().__init__(name=name, **kwargs)
         self.schemes = kwargs.get("schemes", [])
 
-    def validate(self, value, path, **kwargs):
+    def validate(self, value: str | None, path: list, **kwargs: Any) -> str | None:
```
Comment (vegu) [inline src/confu/schema/inet.py:55]: i think that should be path: typing.List[str] @jason20c

File: src/confu/schema/core.py [hunk 6]
```diff
@@ -738,7 +793,7 @@ def validate(schema, config, raise_errors=False, log=None, **kwargs):
         return (success, errors, warnings)
 
 
-def apply_default(config, attribute, path):
+def apply_default(config: dict, attribute: Attribute, path: List[str]) -> None:
```
Comment (vegu) [inline src/confu/schema/core.py:796]: @jason20c hm i think those may need to be `list[str]` since List refers to the confu `List` attribute

Comment (jason20c) [reply src/confu/schema/core.py:796]: Yes I think you're right, will make the changes.

**Review by vegu — APPROVED** [review]

**Review by grizz — APPROVED** [review]

**Pull request merged**